WordPress/gutenberg
# Framework: Add queried-data state helper utility

Related: #6180 
Related: #6360

This pull request seeks to work toward a generalized solution for queried / paginated data resources. It is implemented as an independent (but currently bundled) set of state helpers, including reducer, selectors, and action creators for operating on queried data. Currently, only the categories data has been refactored, and it does not currently take advantage of pagination (as this is not used by the Categories block). Further, as discovered in #6360, some additional work is required for memoization of "same" (but not strictly equal) query fulfillment.

The hope is that each resource (`core-data` state key) can simply reuse the existing helpers from `queried-data`. For terms, this looks like:

```js
export const terms = onSubKey( 'taxonomy' )( queriedDataReducer );
```

Some additional helpers may be required to expand this out to other state like media, particularly for ensuring that the reducer only operates on relevant queried data. For terms this is handled automatically through the `onSubKey` higher-order reducer creator, where state is only tracked for actions which include a `taxonomy` property. Likewise, we may create other higher-order reducers which test for the presence of some resource-specific marker, to be applied as an extension of the baseline `receiveQueriedItems` action creator (e.g. `ifActionMatches( ( action ) => action.queried === 'media' )`).

The state shape is normalized, with queries tracked as stable keys, both to avoid storing redundant data, and to avoid duplicate tracking for equivalent queries:

Example:

```json
{
	"terms": {
		"categories": {
			"items": {
				"1": {
					"id": 1,
					"name": "Uncategorized"
				},
				"2": {
					"id": 2,
					"name": "Cats"
				}
			},
			"queries": {
				"s=cat": {
					"itemIds": [ 1, 2 ],
					"requestingPageByPerPage": {}
				},
				"s=cats": {
					"itemIds": [ 2 ],
					"requestingPageByPerPage": {}
				}
			}
		}
	}
}
```

__Testing instructions:__

This should only impact the Categories block, which is the sole consumer of categories data. Verify that there are no regressions in the display of categories block, including the "Loading" placeholder and ultimately the display of categories.

Ensure unit tests pass:

```
npm test
```
